File: 977-distinct-subsequences-ii/distinct-subsequences-ii.cpp
class Solution {
public:
    int distinctSubseqII(string s) {
        const int MOD = 1e9 + 7;
        vector<int> lastAdded(26, 0);
        long long total = 0; // Total distinct non-empty subsequences so far

        for (char ch : s) {
            int idx = ch - 'a';
            // Distinct subsequences created ending in ch: (total + 1) - lastAdded[idx]
            long long added = (total + 1 - lastAdded[idx] + MOD) % MOD;

            total = (total + added) % MOD;
            lastAdded[idx] = (lastAdded[idx] + added) % MOD;
        }

        return total;
    }
};

File: 977-distinct-subsequences-ii/distinct-subsequences-ii.java
class Solution {
    public int distinctSubseqII(String s) {
        int MOD = 1_000_000_007;
        // end[c] stores the count of distinct subsequences ending with character ('a' + c)
        int[] end = new int[26];
        
        for (char ch : s.toCharArray()) {
            int c = ch - 'a';
            // Current character can append to all existing distinct subsequences,
            // plus form a subsequence of length 1 on its own (+1).
            long sum = 1;
            for (int count : end) {
                sum = (sum + count) % MOD;
            }
            end[c] = (int) sum;
        }
        
        // Sum all subsequences ending with any character
        long ans = 0;
        for (int count : end) {
            ans = (ans + count) % MOD;
        }
        
        return (int) ans;
    }
}